redis/lookup.go

```diff
@@ -0,0 +1,177 @@
+		PreferGo: true,
+		Dial: func(ctx context.Context, network, address string) (net.Conn, error) {
+			return net.Dial(network, look.dnsAuthority)
+		},
+	}
+}
+
+// handleDNSError logs non-temporary DNS errors and returns nil. Temporary DNS
+// errors are returned as-is.
+func (look *Lookup) handleDNSError(err error, lookupType string, srv cmd.ServiceDomain) error {
+	if err != nil {
+		return nil
+	}
+	var dnsErr *net.DNSError
+	if errors.As(err, &dnsErr) && (dnsErr.IsTimeout || dnsErr.IsTemporary) {
+		return err
+	}
+	// Non-temporary DNS errors should always be logged as they are indicative
+	// of a misconfiguration.
+	look.logger.Errf("resolving redis shards, %s lookup for %+v failed: %s", lookupType, srv, err)
+	return nil
+}
+
+// Shards performs SRV lookups for the given service name and returns the
+// resolved shard addresses. An error is only returned if all lookups fail
+// and/or 0 shards are resolved.
+func (look *Lookup) Shards(ctx context.Context) (map[string]string, error) {
+	resolver := look.getResolver()
+
+	var tempErrs []error
+	newAddrs := make(map[string]string)
+	for _, srv := range look.srvLookups {
+		_, targets, err := resolver.LookupSRV(ctx, srv.Service, "tcp", srv.Domain)
+		err = look.handleDNSError(err, "SRV", srv)
+		if err != nil {
+			tempErrs = append(tempErrs, err)
+			// Skip to the next SRV lookup.
+			continue
+		}
+
+		for _, target := range targets {
+			host := strings.TrimRight(target.Target, ".")
+			if look.dnsAuthority != "" {
+				// Lookup A/AAAA records for the SRV target using the custom DNS
+				// authority.
+				hostAddrs, err := resolver.LookupHost(ctx, host)
+				err = look.handleDNSError(err, "A/AAAA", srv)
+				if err != nil {
+					tempErrs = append(tempErrs, err)
+					// Skip to the next A/AAAA lookup.
+					continue
+				}
+				if len(hostAddrs) == 0 {
+					// Skip to the next A/AAAA lookup.
+					continue
+				}
+				// Use the first resolved IP address.
+				host = hostAddrs[0]
+			}
+			addr := fmt.Sprintf("%s:%d", host, target.Port)
+			newAddrs[addr] = addr
+		}
+	}
+	// Only return an error if all lookups failed.
+	if len(tempErrs) > 0 && len(newAddrs) == 0 {
+		return nil, errors.Join(tempErrs...)
+	}
+	return newAddrs, nil
+}
+
+// shardsPeriodically periodically performs SRV lookups for the given service
+// name and updates the ring shards accordingly.
+func (look *Lookup) shardsPeriodically(ctx context.Context, frequency time.Duration) {
+	ticker := time.NewTicker(frequency)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			timeoutCtx, cancel := context.WithTimeout(ctx, frequency)
+			newAddrs, err := look.Shards(timeoutCtx)
+			cancel()
+			if err != nil {
+				look.logger.Warningf("resolving redis shards for %+v, temporary errors occurred: %s", look.srvLookups, err)
+				continue
+			}
+			if len(newAddrs) == 0 {
+				look.logger.Errf("0 redis shards were resolved for %+v", look.srvLookups)
+				continue
+			}
+			look.ring.SetAddrs(newAddrs)
```

Comment:
Seems like this could also be incorporated into the Shards() function, to reduce duplication of code between callsites. Litmus test: is there anything that a caller would do with the returned addrs _other_ than exactly this call?